server/lyft/decorators/audit_project_commands_wrapper.go

```diff
@@ -0,0 +1,142 @@
+package decorators
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/events"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/lyft/aws/sns"
+)
+
+// AuditProjectCommandWrapper is a decorator that notifies sns topic
+// about the state of the command. It is used for auditing purposes
+type AuditProjectCommandWrapper struct {
+	SnsWriter sns.Writer
+	events.ProjectCommandRunner
+}
+
+func (p *AuditProjectCommandWrapper) Apply(ctx models.ProjectCommandContext) models.ProjectResult {
+	id := uuid.New()
+	startTime := strconv.FormatInt(time.Now().Unix(), 10)
+
+	applyEvent := &ApplyEvent{
+		Version:        1,
+		ID:             id.String(),
+		RootName:       ctx.ProjectName,
+		JobType:        models.ApplyCommand,
+		Respository:    ctx.BaseRepo.FullName,
+		Environment:    ctx.Tags["environment"],
+		PullNumber:     ctx.Pull.Num,
+		InitiatingUser: ctx.User.Username,
+		Project:        ctx.Tags["service_name"],
+		ForceApply:     ctx.ForceApply,
+		StartTime:      startTime,
+		Revision:       ctx.Pull.HeadCommit,
+	}
+
+	if err := p.emit(ctx, ApplyEventRunning, applyEvent); err != nil {
+		// return an error if we are not able to write to sns
+		return models.ProjectResult{
+			Error: errors.Wrap(err, "emitting apply event"),
+		}
+	}
+
+	result := p.ProjectCommandRunner.Apply(ctx)
+
+	if result.Error != nil || result.Failure != "" {
+		if err := p.emit(ctx, ApplyEventFailure, applyEvent); err != nil {
+			ctx.Log.Err("failed to emit apply event", err)
+		}
+
+		return result
+	}
+
+	if err := p.emit(ctx, ApplyEventSuccess, applyEvent); err != nil {
+		ctx.Log.Err("failed to emit apply event", err)
+	}
+
+	return result
+}
+
+func (p *AuditProjectCommandWrapper) emit(
+	ctx models.ProjectCommandContext,
+	state EventState,
+	applyEvent *ApplyEvent,
+) error {
+	applyEvent.State = state
+
+	if state == ApplyEventFailure || state == ApplyEventSuccess {
+		applyEvent.EndTime = strconv.FormatInt(time.Now().Unix(), 10)
+	}
+
+	payload, err := applyEvent.Marshal()
+	if err != nil {
+		return errors.Wrap(err, "marshaling apply event")
+	}
+
+	if err := p.SnsWriter.Write(payload); err != nil {
+		return errors.Wrap(err, "writing to sns topic")
+	}
+
+	return nil
+}
+
+// ApplyEvent contains metadata of the state of the apply command
+type ApplyEvent struct {
+	Version        int
+	ID             string
+	State          EventState
+	JobType        models.CommandName
```

Comment:
Think we should create another type for this to derisk in case for some reason models.CommandName changes? It's nice to keep internal types and external types separate.